cmd/sink/expr/filters.go

```diff
@@ -0,0 +1,331 @@
+	f.Lock()
+	defer f.Unlock()
+	maps.Copy(f.archive, namespaceArchive)
+	maps.Copy(f.delete, namespaceDelete)
+	return err
+}
+
+// createFilters returns two maps representing all of the filters for namespace from created by compiling cel
+// expressions from file with matching cel expressions based on GroupVersionKind from globalArchive and globalDelete.
+// The first returned map represents the filters for archiving and the second returned map represents the filters for
+// deleting. createFilters wraps all errors it encounters while creating filters and returns the wrapped error. Even if
+// the returned error is not nil, the maps returned can still be used.
+func (f *Filters) createFilters(
+	namespace string,
+	file string,
+	globalArchive, globalDelete map[string]string,
+) (map[NamespaceGroupVersionKind]cel.Program, map[NamespaceGroupVersionKind]cel.Program, error) {
+	archiveMap := make(map[NamespaceGroupVersionKind]cel.Program)
+	deleteMap := make(map[NamespaceGroupVersionKind]cel.Program)
+	resources, err := k8s.KACResourceSliceFromFile(file)
+	if err != nil {
+		return archiveMap, deleteMap, err
+	}
+	var errList []error
+	for _, resource := range resources {
+		gvk := schema.FromAPIVersionAndKind(resource.Selector.APIVersion, resource.Selector.Kind)
+		namespaceGvk := NamespaceGVKFromNamespaceAndGvk(namespace, gvk)
+		archiveExpr, err := CreateCelExprOr(resource.ArchiveWhen, globalArchive[gvk.String()])
+		if err != nil {
+			errList = append(errList, err)
+		} else {
+			archiveMap[namespaceGvk] = *archiveExpr
+		}
+		deleteExpr, err := CreateCelExprOr(resource.DeleteWhen, globalDelete[gvk.String()])
+		if err != nil {
+			errList = append(errList, err)
+		} else {
+			deleteMap[namespaceGvk] = *deleteExpr
+		}
+	}
+	return archiveMap, deleteMap, errors.Join(errList...)
+}
+
+func (f *Filters) deleteNamespaceFilters(matcher NamespaceMatcher) {
+	f.Lock()
+	defer f.Unlock()
+	maps.DeleteFunc(f.archive, matcher)
+	maps.DeleteFunc(f.delete, matcher)
+}
+
+// Update updates the archive and delete filters when a ConfigMap file changes.
+func (f *Filters) Update(watcher *fsnotify.Watcher) {
+	logger := log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok { // watcher.Close() was called. We will not receive events anymore.
+				return
+			}
+			if files.IsNotDir(event.Name) {
+				f.handleFsEvent(logger, event)
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok { // watcher.Close() was called. We will not receive errors anymore.
+				return
+			}
+			logger.Println("Encountered an error while watching for changes to filters:", err)
+		}
+	}
+}
+
+// handleFsEvent handles the logic for updating filters when an fsnotify.Event is received.
+func (f *Filters) handleFsEvent(logger *log.Logger, event fsnotify.Event) {
+	switch event.Op {
+	case fsnotify.Create:
```

Comment:
Why aren't `fsnotify.Create` and `fsnotify.Write` the same?  I don't think there should be a difference in what is done in both cases.

There may some code from delete that could also be shared.

Reply:
`fsnotify.Create` and `fsnotify.Write` are the if the global filters change.

However, they are different if namespaced filters change. This is because if we receive `fsnotify.Create`, we know that there are not any filters for that namespace so we can just create filters for that namespace and add the them to `filters.archive` and `filters.delete`. If we receive `fsnotify.Write` that means we already have filters for that namespace in `filters.archive` and `filters.delete` therefore, we have to remove the old filters from both maps before we add the new ones. This ensures that only filters that are currently defined in the ConfigMap are used by the sink. I thought about combining them, but unfortunately the logic have to use for deleting the entries requires looping over every key in both `f.archive` and `f.delete`. To speed up the runtime for creating filters for a new namespace, I kept the logic separate

Reply:
You're right, the logic is the same, I'll make a new method that they both call

Reply:
merged the cases for `fsnotify.Create` and `fsnotify.Write`